### PR TITLE
Test package detection in a systematic way

### LINKED ISSUE #39633: Define a way to properly test external detection
# Summary

As a user I want to have a way to test external package detection so to avoid regressions on known use cases.

## Rationale

Currently we have unit-tests to verify compiler detection from known output. As an example:

https://github.com/spack/spack/blob/ecfd9ef12bc4bf39ff92df01921c8f641b54e43e/lib/spack/spack/test/compilers/detection.py#L29-L52

When compiler are moved from internal `Compiler` classes to packages, detection of the compilers on the system will be delegated to external package detection. It is thus necessary to:
1. Define a way to test external package detection
2. Port the current unit-test

to ensure that the new detection mechanism is as reliable as the old one.

## Other references

- https://github.com/spack/spack/pull/18175

### PR DESCRIPTION
**Pull request by alalazo**
closes #21343 
closes #28092 
fixes #39633

This PR adds a new audit sub-command to check that detection of relevant packages is performed correctly in a few scenarios mocking real use-cases. The data for each package being tested is in a YAML file called `detection_test.yaml` alongside the corresponding `package.py` file. 

This is to allow encoding detection tests for compilers and other widely used tools, in preparation for compilers as dependencies.

The audit can be run with:
```console
$ spack -d audit externals 
==> [2023-08-29-15:43:35.481713] Imported audit from built-in commands
==> [2023-08-29-15:43:35.483244] Imported audit from built-in commands
==> [2023-08-29-15:43:35.483774] Reading config from file /home/culpo/PycharmProjects/spack/etc/spack/defaults/repos.yaml
==> [2023-08-29-15:43:35.546046] Reading config from file /home/culpo/PycharmProjects/spack/etc/spack/defaults/config.yaml
==> [2023-08-29-15:43:35.566903] Reading config from file /home/culpo/.spack/config.yaml
==> [2023-08-29-15:43:35.636355] Reading config from file /home/culpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml
==> [2023-08-29-15:43:36.100078] '/tmp/tmpmvgxs0xq/bin/g++' '--version'
==> [2023-08-29-15:43:36.101754] '/tmp/tmpmvgxs0xq/bin/g++' '-dumpfullversion'
==> [2023-08-29-15:43:36.103053] '/tmp/tmpmvgxs0xq/bin/gcc' '--version'
==> [2023-08-29-15:43:36.104154] '/tmp/tmpmvgxs0xq/bin/gcc' '-dumpfullversion'
==> [2023-08-29-15:43:36.108037] '/tmp/tmpwfh68a7_/bin/g++-5' '--version'
==> [2023-08-29-15:43:36.109346] '/tmp/tmpwfh68a7_/bin/g++-5' '-dumpfullversion'
==> [2023-08-29-15:43:36.110537] '/tmp/tmpwfh68a7_/bin/g++-5' '-dumpversion'
==> [2023-08-29-15:43:36.111552] '/tmp/tmpwfh68a7_/bin/gcc-5' '--version'
==> [2023-08-29-15:43:36.112801] '/tmp/tmpwfh68a7_/bin/gcc-5' '-dumpfullversion'
==> [2023-08-29-15:43:36.113896] '/tmp/tmpwfh68a7_/bin/gcc-5' '-dumpversion'
==> [2023-08-29-15:43:36.114965] '/tmp/tmpwfh68a7_/bin/gfortran-5' '--version'
==> [2023-08-29-15:43:36.115915] '/tmp/tmpwfh68a7_/bin/gfortran-5' '-dumpfullversion'
==> [2023-08-29-15:43:36.116910] '/tmp/tmpwfh68a7_/bin/gfortran-5' '-dumpversion'
==> [2023-08-29-15:43:36.121359] '/tmp/tmp2_x568wi/bin/x86_64-linux-gnu-gcc-10' '--version'
==> [2023-08-29-15:43:36.122565] '/tmp/tmp2_x568wi/bin/x86_64-linux-gnu-gcc-10' '-dumpfullversion'
==> [2023-08-29-15:43:36.123825] '/tmp/tmp2_x568wi/bin/x86_64-linux-gnu-gcc-6' '--version'
==> [2023-08-29-15:43:36.124894] '/tmp/tmp2_x568wi/bin/x86_64-linux-gnu-gcc-6' '-dumpfullversion'
==> [2023-08-29-15:43:36.125865] '/tmp/tmp2_x568wi/bin/x86_64-linux-gnu-g++-10' '--version'
==> [2023-08-29-15:43:36.127096] '/tmp/tmp2_x568wi/bin/x86_64-linux-gnu-g++-10' '-dumpfullversion'
==> [2023-08-29-15:43:36.138709] '/tmp/tmp1w5upxqk/bin/intel64/icpc' '--version'
==> [2023-08-29-15:43:36.140950] '/tmp/tmp1w5upxqk/bin/intel64/ifort' '--version'
==> [2023-08-29-15:43:36.142114] '/tmp/tmp1w5upxqk/bin/intel64/icc' '--version'
==> [2023-08-29-15:43:36.206720] '/tmp/tmpkujzcakc/bin/clang-3.9' '--version'
==> [2023-08-29-15:43:36.208211] '/tmp/tmpkujzcakc/bin/clang++-3.9' '--version'
==> [2023-08-29-15:43:36.213159] '/tmp/tmpv5s6fkek/bin/clang++-8' '--version'
==> [2023-08-29-15:43:36.214300] '/tmp/tmpv5s6fkek/bin/clang++-3.9' '--version'
==> [2023-08-29-15:43:36.215466] '/tmp/tmpv5s6fkek/bin/clang-3.9' '--version'
==> [2023-08-29-15:43:36.216532] '/tmp/tmpv5s6fkek/bin/clang-8' '--version'
==> [2023-08-29-15:43:36.217426] '/tmp/tmpv5s6fkek/bin/ld.lld-8' '--version'
==> [2023-08-29-15:43:36.218345] '/tmp/tmpv5s6fkek/bin/lldb' '--version'
==> [2023-08-29-15:43:36.222536] '/tmp/tmpl0yvyr1h/bin/clang++' '--version'
==> [2023-08-29-15:43:36.224217] '/tmp/tmpl0yvyr1h/bin/clang' '--version'
==> [2023-08-29-15:43:36.225956] The following executables in /tmp/tmpl0yvyr1h/bin were decidedly not part of the package llvm: /tmp/tmpl0yvyr1h/bin/clang++, /tmp/tmpl0yvyr1h/bin/clang
PKG-EXTERNALS: 0 issues found.
```

Modifications:

- [x] Added an audit to check that detection for packages is performed correctly
- [x] Added documentation on how to write detection tests for each package
- [x] Added detection tests for `gcc`, `llvm` and `intel`
- [x] Run detection audits in CI

### CODE REVIEW

File: lib/spack/spack/test/cmd/external.py [hunk 1]
```diff
@@ -240,3 +244,71 @@ def test_new_entries_are_reported_correctly(
...
+    all_packages = spack.repo.path.all_packages()
+    to_be_tested = []
+    for pkg in all_packages:
+        pkg_dir = os.path.dirname(
+            spack.repo.path.filename_for_package_name(pkg.name)
+        )
+        detection_data = os.path.join(pkg_dir, 'detection_test.yaml')
+        if os.path.exists(detection_data):
+            to_be_tested.append(pkg.name)
+    return to_be_tested
```
Comment (alalazo) [inline lib/spack/spack/test/cmd/external.py]: Note that this will slow-down start-up of unit tests, but at the moment I can't think of a better way to retrieve this information.

File: lib/spack/spack/cmd/external.py [hunk 2]
```diff
@@ -185,12 +185,16 @@ def _convert_to_iterable(single_val_or_multiple):
 
 
 def _determine_base_dir(prefix):
-    # Given a prefix where an executable is found, assuming that prefix ends
-    # with /bin/, strip off the 'bin' directory to get a Spack-compatible
+    # Given a prefix where an executable is found, assuming that prefix
+    # contains /bin/, strip off the 'bin' directory to get a Spack-compatible
```
Comment (scheibelp) [inline lib/spack/spack/cmd/external.py]: > strip off the bin directory

... and all subsequent directories

File: lib/spack/spack/test/cmd/external.py (hunk 1 above)
Comment (scheibelp) [reply lib/spack/spack/test/cmd/external.py]: Wouldn't this fixture not be run until the unit test that needs it (i.e. I think this wouldn't run until the test that needs it is run)? We have some long-running fixtures so this seems fine to me either way.

It might be faster to do some equivalent of Linux `find . -name "detection_test.yaml"` since that would skip importing of modules, but I don't think this needs to be particularly fast.

File: lib/spack/docs/packaging_guide.rst [hunk 3]
```diff
@@ -4317,6 +4317,86 @@ follows:
...
+Add detection tests to packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To ensure that a software is detected correctly for multiple configurations
+and on different systems users can write a ``detection_test.yaml`` file and
+put it in the same directory as the corresponding ``package.py`` file.
+This YAML file contains enough information for Spack to mock an environment
+and try to check if the detection logic yields the results that are expected.
+
+As a general rule, attributes at the top-level of ``detection_test.yaml``
```
Comment (scheibelp) [inline lib/spack/docs/packaging_guide.rst:6209]: I'm concerned about having a domain-specific language vs. adding functions to the package itself -  why not have the user write a function in `package.py`?

```
def test_inspections(self):
  bin1 = 'bin/clang-3.9'
  script1 = """\
echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
echo "Target: x86_64-pc-linux-gnu"
echo "Thread model: posix"
echo "InstalledDir: /usr/bin"
"""
  bin2 = ...
  result2 = ...

  specs = ['llvm@3.9.1 +clang~lld~lldb']

  return [(bin1, result1), (bin2, result2)], specs
```

File: lib/spack/docs/packaging_guide.rst [hunk 4]
```diff
@@ -4317,6 +4317,86 @@ follows:
...
+   :header-rows: 1
+
+   * - Option Name
+     - Description
+     - Allowed Values
+     - Required
+   * - ``layout``
+     - Specifies the filesystem tree used for the test
+     - List of objects
+     - Yes
```
Comment (scheibelp) [inline lib/spack/docs/packaging_guide.rst:6259]: What does "Yes" mean here?

**Review by scheibelp — DISMISSED** [review]
I have a few questions. Also I think a function in `package.py` would be easier to use than a domain specific language. More details in the comments.

File: lib/spack/docs/packaging_guide.rst (hunk 3 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst:6209]: I don't see this as a domain specific language, more like an additional YAML file that packagers can provide to specify tests for package detection and aggregate knowledge on different cases. Given that:
- Our users in general need to deal with YAML files a lot (so they are used to them)
- This particular file is not mandatory and is documented

I wouldn't consider this much of an issue. Maybe there's a better section to put documentation on it rather than the one I used?

A few considerations. The YAML file has no behavior, only static information to setup a unit test on a core function of the:
```console
$ spack external find ...
``` 
command. The logic for this unit test is the same for all packages. In some sense the information in the YAML file is of the same kind of what we can see in a parametrized fixture for other tests. The only difference is that here this information relates to different packages and is more verbose, so I decided to put it in a YAML file alongside the corresponding `package.py` rather than aggregating it in a single test file.

There's also a thing that worries me in taking the approach of an additional method. The package class is already overcharged with a lot of different responsibilities and I don't think it would help readability or maintenance to also add the responsibility of unit testing its own detection.

File: lib/spack/docs/packaging_guide.rst (hunk 4 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst:6259]: This part renders in the docs as a table. It describes in details the attributes that can be added to the YAML file:

Screenshot from 2020-08-24 22-02-46

The "Yes" is under the "Required" column, to mean that the field is required. It's in fact a bit silly that now every field is required, but this was done thinking of future extensions to e.g. check that an external attribute is computed in a certain way. That wouldn't be required since not all packages have external attributes. If you think it's of no use at the moment I can remove the entire column.

File: lib/spack/spack/test/cmd/external.py (hunk 1 above)
Comment (alalazo) [reply lib/spack/spack/test/cmd/external.py]: This is a function to initialize a fixture, so it's run when tests are collected not when they are run - which is not optimal. I'll take your suggestion of not going through the import mechanism and see if that improves the current situation.

Comment (alalazo) [reply lib/spack/spack/test/cmd/external.py]: Implemented your suggestion and used `glob.glob`. Brilliant, it's so much faster! :tada:

File: lib/spack/docs/packaging_guide.rst (hunk 3 above)
Comment (scheibelp) [reply lib/spack/docs/packaging_guide.rst:6209]: > I don't see this as a domain specific language, more like an additional YAML file that packagers can provide

> Our users in general need to deal with YAML files a lot (so they are used to them)

I think what you are asking is: “if we have config files, then why do you think this is bad?” Overall I would say the problem is that this doesn’t provide any of the benefits that config files typically provide to Spack users:

Users that write `detection_test.yaml` have to be familiar with the detection API and the specific implementation of the detection function in the associated `package.py` file. Users editing config files don’t need to know Spack internals, but users writing `detection_test.yaml` do.

Generally we keep config files as key/value stores or otherwise make them as simple as possible. Even the most complex config files we have in spack are entries which are collections of properties. `detection_test.yaml` specifies properties (like file locations) along with behavior (scripts) and outcomes (the expected spec output): that is more-involved than any config file we have. This also forces users to be familar with YAML syntax constraints which don’t appear in other config files, like the text block specification:

```
output: |
   echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
   echo "Target: x86_64-pc-linux-gnu"
   echo "Thread model: posix"
   echo "InstalledDir: /usr/bin"
```

Since the user has to understand (and in fact, coordinate with) the internals (i.e. the external detection function in `package.py`), I think this behavior should be specified as a function rather than a YAML file.

Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst:6209]: > I think what you are asking is: “if we have config files, then why do you think this is bad?” Overall I would say the problem is that this doesn’t provide any of the benefits that config files typically provide to Spack users:

That's not the point. As I tried to explain above this YAML is not configuration - it's data fed into a unit test. I concur this part is true:

> Users that write detection_test.yaml have to be familiar with the detection API and the specific implementation of the detection function in the associated package.py file.

even though I don't see that as an issue, but then I don't understand the alternative solution of adding a method to the package. It shares the same charateristics (be familiar with the detection API and specific implementation of other functions in the package) AND it couples the package class with unit tests on it. Overall it seems to me no concern would be solved but more would be added.

**TL;DR** I pushed a modification in 55335df that moves the YAML files into a subfolder of the directory where we keep other unit test data files and moved the docs to the developer section. I hope this could be a good way to mitigate the concern you have over the user facing side of this PR and the one I have over adding an additional function to packages and coupling them to unit tests. The description at the top of the PR has been updated too.

Comment (scheibelp) [reply lib/spack/docs/packaging_guide.rst:6209]: > Overall it seems to me no concern would be solved but more would be added.

I want to avoid forcing users to switch between paradigms: if a user needs to be familiar with the external detection implementation of a package, then they will be reading Python. I don't want to force them to switch between Python and a YAML-based specification of behavior - I think they might as well write a Python function.

> I don't understand the alternative solution of adding a method to the package. It shares the same characteristics (be familiar with the detection API and specific implementation of other functions in the package)

> AND it couples the package class with unit tests on it

I think that is a good thing. They are already coupled, I don't see why separating them into different files changes that.

Put another way: what do you think is wrong with placing the testing logic in the package?

Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst:6209]: To me the data in the YAML file is of the same kind as this:

https://github.com/spack/spack/blob/77a28b81ac10d63bca30c66edfd0205a11a1235e/lib/spack/spack/test/compilers/detection.py#L27-L38

so I'll counter a question with a question: do you think we should move that into the corresponding compiler class to be returned from a method:

```python
class Arm(Compiler):
    def input_for_version_detection(self):
        return [
            ('Arm C/C++/Fortran Compiler version 19.0 (build number 73) (based on LLVM 7.0.2)\n' # NOQA 
             'Target: aarch64--linux-gnu\n' 
             'Thread model: posix\n' 
             'InstalledDir:\n' 
             '/opt/arm/arm-hpc-compiler-19.0_Generic-AArch64_RHEL-7_aarch64-linux/bin\n', # NOQA 
             '19.0.0.73'), 
            ('Arm C/C++/Fortran Compiler version 19.3.1 (build number 75) (based on LLVM 7.0.2)\n' # NOQA 
             'Target: aarch64--linux-gnu\n' 
             'Thread model: posix\n' 
             'InstalledDir:\n' 
             '/opt/arm/arm-hpc-compiler-19.0_Generic-AArch64_RHEL-7_aarch64-linux/bin\n', # NOQA 
             '19.3.1.75') ]
```
?

Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst:6209]: > what do you think is wrong with placing the testing logic in the package?

In the current form the unit test has knowledge of:
- The behavior of `spack external find` (or functions called within that module)
- The data needed to mock a filesystem structure and the expected result of a search

and drives the test - the package class doesn't know about the test, as it's not its responsibility.

With a method in the package class we have a unit test that calls a package that knows how the unit test is structured and returns the appropriate input - the unit test knows about the package and the package knows about the unit test and this makes the entire structure more coupled.

Being that 55335df makes it clear that the data is for unit tests and moves it away from packages, would that be good enough for you?

**Comment by tldahlgren** [comment]
Closing and re-opening this PR to see if doing so will result in reporting the status of the three outstanding checks.

File: lib/spack/docs/packaging_guide.rst (hunk 3 above)
Comment (scheibelp) [reply lib/spack/docs/packaging_guide.rst:6209]: > so I'll counter a question with a question: do you think we should move that into the corresponding compiler class to be returned from a method:

Yes: all the compiler classes will eventually be packages anyways, so it would fit my expectation to move the example output into the compiler class (in the same way that I want to move the example package detection output into the package class).

> With a method in the package class we have a unit test that calls a package that knows how the unit test is structured and returns the appropriate input - the unit test knows about the package and the package knows about the unit test and this makes the entire structure more coupled.

I disagree with the first point of this and therefore with the conclusion, specifically you say:

> With a method in the package class we have a unit test that calls a package that knows how the unit test is structured and returns the appropriate input

Just because the package would need to follow a specific API to return example version output doesn't mean that it needs to know about the structure of the unit test. The API it needs to follow is:

```
class ExampleOutput(object):
  def __init__(self):
    self.expected_version = '1.0
    self.exe_to_output = {
      'bin/exe1': 'version 1.0',
      'bin/exe2': '1.0'
      ...
    }

class Package(object):
  def version_examples(self):
    # Return a set of ExampleOutput objects
   return []
```

But it wouldn't need to know how the unit test works

**Comment by becker33** [comment]
I prefer this PR to the alternative implementations

File: lib/spack/spack/test/detection.py [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+@pytest.mark.detection
+@pytest.mark.parametrize('package_name', candidate_packages())
+def test_package_detection(mock_executable, package_name):
+    def detection_tests_for(pkg):
+        pkg_dir = os.path.dirname(
+            spack.repo.path.filename_for_package_name(pkg)
+        )
+        detection_data = os.path.join(pkg_dir, 'detection_test.yaml')
+        with open(detection_data) as f:
+            return spack.util.spack_yaml.load(f)
```
Comment (becker33) [inline lib/spack/spack/test/detection.py]: This should move outside of this test, there's no reason for it to be nested

File: lib/spack/spack/test/detection.py [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+            exe = mock_executable(
+                binary['name'], binary['output'], subdir=binary['subdir']
+            )
+            to_be_removed.append(exe)
+            hints.add(os.path.dirname(str(exe)))
+
+        yield list(hints)
+
+        for exe in to_be_removed:
+            os.unlink(exe)
```
Comment (becker33) [inline lib/spack/spack/test/detection.py]: I think this can move out from being nested as well, but it will require some changes to how it calls mock_executable, so that might not be worth it.

File: lib/spack/spack/test/detection.py [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+            return result
+
+        monkeypatch.setattr(spack.detection.path, 'executables_in_path', _mock_search)
+    return _factory
+
+
+def test_find_external_single_package(mock_executable, executables_found):
+    pkgs_to_check = [spack.repo.get('cmake')]
+    executables_found({
+        mock_executable("cmake", output='echo "cmake version 1.foo"'): 'cmake'
```
Comment (becker33) [inline lib/spack/spack/test/detection.py]: This PR might not be the place to fix this, but the `output` kwarg to `mock_executable` either needs to be changed to be `contents`, or the fixture needs to be changed to automatically prepend the `echo` portion.

File: var/spack/repos/builtin/packages/gcc/detection_test.yaml [hunk 8]
```diff
@@ -0,0 +1,46 @@
+paths:
+  # Ubuntu 18.04, system compilers without Fortran
+  - layout:
+      - subdir: [bin]
+        name: gcc
+        output: "echo 7.5.0"
```
Comment (becker33) [inline var/spack/repos/builtin/packages/gcc/detection_test.yaml]: Can we make this take `"7.5.0"` and have the internal logic figure out how to make it print that output? `echo` isn't part of the output we care about at all.

File: lib/spack/spack/test/cmd/external.py [hunk 9]
```diff
@@ -10,86 +10,10 @@
 import spack
 import spack.detection
 import spack.detection.path
-from spack.main import SpackCommand
-from spack.spec import Spec
+import spack.main
+import spack.util.spack_yaml
```
Comment (becker33) [inline lib/spack/spack/test/cmd/external.py]: This appears not to be used anywhere

File: lib/spack/docs/packaging_guide.rst [hunk 10]
```diff
@@ -5388,6 +5388,86 @@ follows:
 
 .. _package-lifecycle:
 
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Add detection tests to packages
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+To ensure that a software is detected correctly for multiple configurations
```
Comment (becker33) [inline lib/spack/docs/packaging_guide.rst]:
```suggestion
To ensure that software is detected correctly for multiple configurations
```

File: lib/spack/docs/packaging_guide.rst [hunk 11]
```diff
@@ -5388,6 +5388,86 @@ follows:
...
+and on different systems users can write a ``detection_test.yaml`` file and
+put it in the same directory as the corresponding ``package.py`` file.
```
Comment (becker33) [inline lib/spack/docs/packaging_guide.rst]:
```suggestion
put it in the packaage directory alongside the ``package.py`` file.
```

File: lib/spack/docs/packaging_guide.rst [hunk 12]
```diff
@@ -5388,6 +5388,86 @@ follows:
...
+This YAML file contains enough information for Spack to mock an environment
+and try to check if the detection logic yields the results that are expected.
+
+As a general rule, attributes at the top-level of ``detection_test.yaml``
+represent search mechanisms and they all map to a list of tests that should confirm
```
Comment (becker33) [inline lib/spack/docs/packaging_guide.rst]:
```suggestion
represent search mechanisms and they each map to a list of tests that should confirm
```

File: lib/spack/docs/packaging_guide.rst [hunk 13]
```diff
@@ -5388,6 +5388,86 @@ follows:
...
+the validity of each package detection logic.
```
Comment (becker33) [inline lib/spack/docs/packaging_guide.rst]:
```suggestion
the validity of the package's detection logic.
```

File: lib/spack/docs/packaging_guide.rst [hunk 14]
```diff
@@ -5388,6 +5388,86 @@ follows:
 
 .. _package-lifecycle:
```
Comment (becker33) [inline lib/spack/docs/packaging_guide.rst]: I think this may be in the wrong place now

File: lib/spack/docs/packaging_guide.rst [hunk 15]
```diff
@@ -5388,6 +5388,86 @@ follows:
...
+This YAML file contains enough information for Spack to mock an environment
+and try to check if the detection logic yields the results that are expected.
+
+As a general rule, attributes at the top-level of ``detection_test.yaml``
+represent search mechanisms and they all map to a list of tests that should confirm
+the validity of each package detection logic.
+
+""""""""""""""""""""""""""
+Tests for PATH inspections
+""""""""""""""""""""""""""
```
Comment (becker33) [inline lib/spack/docs/packaging_guide.rst:6223]: Do we anticipate any other package detection method that we should be including tests for? We have the module-based detection for cray, but (a) I don't think we currently test it, because we can't test on the cray PE, and (b) I think we plan to deprecate that support as soon as we possibly can.

File: lib/spack/docs/packaging_guide.rst [hunk 16]
```diff
@@ -5388,6 +5388,86 @@ follows:
...
+         echo "InstalledDir: /usr/bin"
+     - subdir: [bin]
+       name: clang++-3.9
+       output: |
+         echo "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
+         echo "Target: x86_64-pc-linux-gnu"
+         echo "Thread model: posix"
+         echo "InstalledDir: /usr/bin"
+     results:
+     - spec: 'llvm@3.9.1 +clang~lld~lldb'
```
Comment (becker33) [inline lib/spack/docs/packaging_guide.rst]: I'm worried that this format is very obtuse. Would we lose any information if we switched to something like:

```suggestion
   - layout:
     - executable: bin/clang-3.9
       output: |
        "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
        "Target: x86_64-pc-linux-gnu"
        "Thread model: posix"
        "InstalledDir: /usr/bin"
     - executable: bin/clang++-3.9
       output: |
        "clang version 3.9.1-19ubuntu1 (tags/RELEASE_391/rc2)"
        "Target: x86_64-pc-linux-gnu"
        "Thread model: posix"
        "InstalledDir: /usr/bin"
     specs:
     - 'llvm@3.9.1 +clang~lld~lldb'
```

This would seem to remove unnecessary verbosity.

**Review by becker33 — CHANGES_REQUESTED** [review]
@alalazo sorry about the delay in getting this reviewed.

File: lib/spack/docs/packaging_guide.rst (hunk 15 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst:6223]: Fwiw, we still have module based search for Cray and we might add other inspection method too - e.g. maybe querying system package managers etc.

File: lib/spack/docs/packaging_guide.rst (hunk 16 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst]: Done in https://github.com/spack/spack/pull/18175/commits/d40f73f6d71e9a1583411d306d957a264f50a05b

File: lib/spack/docs/packaging_guide.rst (hunk 13 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst]: Done in https://github.com/spack/spack/pull/18175/commits/d1c916b50af4517a2ee0b09f794fabceaedf3e2d

File: lib/spack/docs/packaging_guide.rst (hunk 12 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst]: Done in https://github.com/spack/spack/pull/18175/commits/d1c916b50af4517a2ee0b09f794fabceaedf3e2d

File: lib/spack/docs/packaging_guide.rst (hunk 10 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst]: Done in https://github.com/spack/spack/pull/18175/commits/d1c916b50af4517a2ee0b09f794fabceaedf3e2d

File: lib/spack/docs/packaging_guide.rst (hunk 11 above)
Comment (alalazo) [reply lib/spack/docs/packaging_guide.rst]: Done in https://github.com/spack/spack/pull/18175/commits/d1c916b50af4517a2ee0b09f794fabceaedf3e2d

File: lib/spack/spack/test/detection.py (hunk 5 above)
Comment (alalazo) [reply lib/spack/spack/test/detection.py]: Done

File: lib/spack/spack/test/detection.py (hunk 7 above)
Comment (alalazo) [reply lib/spack/spack/test/detection.py]: Outdated

File: lib/spack/spack/test/detection.py (hunk 6 above)
Comment (alalazo) [reply lib/spack/spack/test/detection.py]: Outdated

File: lib/spack/spack/test/cmd/external.py (hunk 9 above)
Comment (alalazo) [reply lib/spack/spack/test/cmd/external.py]: Outdated

File: var/spack/repos/builtin/packages/gcc/detection_test.yaml (hunk 8 above)
Comment (alalazo) [reply var/spack/repos/builtin/packages/gcc/detection_test.yaml]: There are cases where we want to mock some specific behavior from an executable. For instance, a `gcc` that fails with an error message and an exitcode if called with `--version`, but returns some output if called with `-dumpversions`. To deal with that I renamed `output` to `script`, since that seems more appropriate given the nature of that field.

File: lib/spack/spack/audit.py [hunk 17]
```diff
@@ -798,3 +801,75 @@ def _analyze_variants_in_directive(pkg, constraint, directive, error_cls):
...
+    repo_dirs = [spack.util.path.canonicalize_path(x) for x in spack.config.get("repos")]
+
+    # Compute which files need to be tested
+    to_be_tested = []
+    for repo_dir in repo_dirs:
+        pattern = os.path.join(repo_dir, "packages", "**", "detection_test.yaml")
+        pkgs_with_tests = [os.path.basename(os.path.dirname(x)) for x in glob.glob(pattern)]
+        to_be_tested.extend(pkgs_with_tests)
+
+    return to_be_tested
```
Comment (becker33) [inline lib/spack/spack/audit.py]: I think this will incorrectly flag a package as having a detection_test.yaml if the package is overridden in a higher priority repo and the overriding package does not inherit from the one with the tests (as the tests will now be invalid).

I think a possible solution to this would be to add a PackageBase instance method `get_detection_test_info`. This method would:

1. check for a `detection_tests.yaml` in the current directory
2. check any superclass in MRO below `PackageBase` for a detection_tests.yaml` file
3. return any files found (so tests could be layered between base and inheritor packages)

File: lib/spack/spack/detection/test.py [hunk 18]
```diff
@@ -0,0 +1,142 @@
...
+
+    This function reads that file to create a bunch of ``Runner`` objects.
+
+    Args:
+        pkg_name: name of the package to test
+        repository: repository where the package lives
+    """
+    result = []
+    pkg_dir = pathlib.Path(repository.filename_for_package_name(pkg_name)).parent
+    detection_tests_yaml = pkg_dir / "detection_test.yaml"
```
Comment (becker33) [inline lib/spack/spack/detection/test.py]: This could be improved by my recommendation above re: an instance method to find detection test files

File: lib/spack/spack/audit.py [hunk 19]
```diff
@@ -798,3 +801,75 @@ def _analyze_variants_in_directive(pkg, constraint, directive, error_cls):
...
+            spack.detection.detection_tests(pkg_name, spack.repo.PATH)
+        ):
+            specs = test_runner.execute()
+            expected_specs = test_runner.expected_specs
+
+            if not expected_specs and specs:
+                summary = pkg_name + ": detected a spec when expecting none"
+                details = [f'"{s}" was detected [test_id={idx}]' for s in specs]
+                errors.append(error_cls(summary=summary, details=details))
+
```
Comment (becker33) [inline lib/spack/spack/audit.py]: This block seems unnecessary considering we have good output for the case `set(specs) - set(expected_specs)` is non-empty.

File: lib/spack/spack/detection/test.py [hunk 20]
```diff
@@ -0,0 +1,142 @@
...
+            specs = set(x.spec for x in entries[self.test.pkg_name])
+        return list(specs)
+
+    @contextlib.contextmanager
+    def _mock_layout(self) -> Generator[List[str], None, None]:
+        hints, to_be_removed = set(), []
+        try:
+            for entry in self.test.layout:
+                exes = self._create_executable_scripts(entry)
+                to_be_removed.extend(exes)
```
Comment (becker33) [inline lib/spack/spack/detection/test.py]: I think this could leave files behind if `self._create_executable_scripts(entry)` raises an error. We should refactor this so the list of files to remove is calculated first, then the attempt to create them. It probably costs an extra iteration over the set of tests, but the list isn't long enough for that to matter compared to the I/O times.

**Review by becker33 — CHANGES_REQUESTED** [review]
Question: Is your plan to extend this to test `extra_attributes` in a later PR?

Otherwise mostly looks good, change requests noted inline

File: lib/spack/spack/audit.py (hunk 17 above)
Comment (alalazo) [reply lib/spack/spack/audit.py]: There's a _small_ performance issue :slightly_smiling_face: If I cycle over packages and read them things can get extremely slow. 

Below, on my laptop, that means going from `0.8s` of the current implementation to `28s`:
Screenshot from 2023-09-04 10-42-00

So this is roughly a 35x slowdown :thinking: Let me think if there's a faster solution to this.

Comment (alalazo) [reply lib/spack/spack/audit.py]: See https://github.com/spack/spack/pull/18175/commits/b9c42ab5bee884f0bdaad1a911c57d9d650b69af This requires to add a file like:
```yaml
includes:
- "builtin.gcc"
```
to reuse the detection tests file for a custom GCC, but still runs in `0.8s` instead of half a minute.

File: lib/spack/spack/detection/test.py (hunk 18 above)
Comment (alalazo) [reply lib/spack/spack/detection/test.py]: See https://github.com/spack/spack/pull/18175#discussion_r1314988870

File: lib/spack/spack/audit.py (hunk 19 above)
Comment (alalazo) [reply lib/spack/spack/audit.py]: Fixed in https://github.com/spack/spack/pull/18175/commits/1ba5ec0998259d4c1fc707ec16436122645585f1

File: lib/spack/spack/detection/test.py (hunk 20 above)
Comment (alalazo) [reply lib/spack/spack/detection/test.py]: TIL `tempfile.TemporaryDirectory` https://github.com/spack/spack/pull/18175/commits/fbf335c8b1644cc315e8916990c6cf69c32a1c70

**Comment by alalazo** [comment]
> Is your plan to extend this to test extra_attributes in a later PR?

Yes, that would be the idea. I'd like to check which extra attributes we need besides `cc` etc. and see how they can be tested.

**Comment by alalazo** [comment]
@becker33

**Comment by alalazo** [comment]
@becker33 How should we proceed with this? If there are doubts on `includes:` we can either:
1. Document it as experimental, saying that the feature might change at any time
2. Leave it out of this PR, and document the limitation that external tests read the YAML file in the same directory as `package.py`

I think, at least initially, the percentage of people who'll add a detection test in their repo which overrides a detection test in built-in will be next to zero - so I wouldn't block this PR on that.

**Review by becker33 — APPROVED** [review]
I still think that we'd be better off using a package attribute and the time savings isn't relevant compared to modeling it properly, but I recognize I'm being outvoted on this issue.
